File: gunhoflash/validate-binary-search-tree.py
import sys

class TreeNode:
  def __init__(self, val = 0, left = None, right = None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def checkBST(self, root: Optional[TreeNode], minVal: int, maxVal: int) -> bool:
    if root is None: return True
    if root.val <= minVal or root.val >= maxVal: return False
    return self.checkBST(root.left, minVal, root.val) and self.checkBST(root.right, root.val, maxVal)

  def isValidBST(self, root: Optional[TreeNode]) -> bool:
    return self.checkBST(root, float('-inf'), float('inf'))

File: gunhoflash/validate-binary-search-tree.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function isValidBST(root: TreeNode | null): boolean {
  function checkBST(root: TreeNode | null, min: number, max: number): boolean {
    if (root === null) return true;
    if (root.val <= min || root.val >= max) return false;
    return checkBST(root.left, min, root.val) && checkBST(root.right, root.val, max);
  }
  return checkBST(root, -Infinity, Infinity);
}
